app/src/org/commcare/android/javarosa/IntentCallout.java

```diff
@@ -207,28 +213,44 @@ private boolean processOdkResponse(Intent intent, TreeReference intentQuestionRe
         // see if we have a return bundle
         Bundle response = intent.getBundleExtra(INTENT_RESULT_EXTRAS_BUNDLE);
 
-        // Load all of the data from the incoming bundle
-        if (responseToRefMap != null && response != null) {
-            for (String key : responseToRefMap.keySet()) {
-                // See if the value exists at all, if not, skip it
-                if (!response.containsKey(key)) {
-                    continue;
+        if (responseToRefMap != null) {
+            if (response != null) {
+                // Load all of the data from the incoming bundle
+                for (String key : responseToRefMap.keySet()) {
+                    // See if the value exists at all, if not, skip it
+                    if (!response.containsKey(key)) {
+                        continue;
+                    }
+                    setOdkResponseValue(intentQuestionRef, destination, response.getString(key), responseToRefMap.get(key));
                 }
-
-                // Get our response value
-                String responseValue = response.getString(key);
-                if (responseValue == null) {
-                    responseValue = "";
+            } else {
+                // Check if intent has response keys as extras.
+                boolean hasExtra = false;
+                for (String key: responseToRefMap.keySet()) {
+                    if (!intent.hasExtra(key)) {
+                        continue;
+                    }
+                    hasExtra = true;
+                    setOdkResponseValue(intentQuestionRef, destination, intent.getStringExtra(key), responseToRefMap.get(key));
                 }
-
-                for (TreeReference ref : responseToRefMap.get(key)) {
-                    processResponseItem(ref, responseValue, intentQuestionRef, destination);
+                if (hasExtra) {
+                    return true;
                 }
             }
         }
         return (result != null);
     }
 
+    private void setOdkResponseValue(TreeReference intentQuestionRef, File destination, String responseValue, Vector<TreeReference> responseRefs) {
+        if (responseValue == null) {
```

Comment:
I think a better approach would be just to skip the call to `processResponseItem` when `responseValue` is null.

Reply:
The earlier code uses empty value when `responseValue` is null. If that needs to be changed, then perhaps we can `return` here when the value is null?

Reply:
Ahh didn't realize it's like this before, no need for a change in behaviour in that case.